dashboard/app/models/lesson.rb

```diff
@@ -89,6 +89,36 @@ class Lesson < ApplicationRecord
 
   include CodespanOnlyMarkdownHelper
 
+  def self.update_lessons_in_migrated_unit(unit, lesson_group, raw_lessons, counters)
+    raw_lessons.map do |raw_lesson|
+      lesson = fetch_lesson(raw_lesson, unit)
+
+      numbered_lesson = !!raw_lesson[:has_lesson_plan] || !raw_lesson[:lockable]
+
+      lesson.assign_attributes(
+        lesson_group: lesson_group,
+        absolute_position: (counters.lesson_position += 1),
+        relative_position: numbered_lesson ? (counters.numbered_lesson_count += 1) : (counters.unnumbered_lesson_count += 1)
+      )
+      lesson.save! if lesson.changed?
+      lesson
+    end
+  end
+
+  def self.fetch_lesson(raw_lesson, unit)
+    if raw_lesson[:id]
+      return Lesson.find_by!(script: unit, id: raw_lesson[:id], key: raw_lesson[:key])
+    end
+    Lesson.prevent_blank_display_name(raw_lesson)
+    Lesson.create!(
+      key: raw_lesson[:key],
+      script: unit,
+      name: raw_lesson[:name],
+      relative_position: 0,  # will be updated by the caller, but can't be nil
+      has_lesson_plan: true
```

Comment:
Should this be set based on `raw_lesson`?

Reply:
when creating a lesson from the unit edit page, there is no way to specify that the lesson should have `has_lesson_plan: false`, so I didn't see a reason to accept this param from the client.

Reply:
Ah interesting. I figured it was there because it's used on line 96 above. Maybe that line should use the lesson object instead of `raw_lesson`?

Reply:
ah, yes, that seems better!